url7f15a160d2_timofei7_old_ai_assignments_tgz-pJ8-org_json_TestJ8/src/hw5/Words.java

```diff
@@ -0,0 +1,282 @@
+			if (accum >= dart)
+			{
+				//System.out.println(bigramCondProb(s.string, bigram));
+				return s.string;
+			}
+		}
+		
+		return "";
+	}
+	
+	
+	/**
+	 * builds a 15 word sentence
+	 * @param bigram
+	 * @return
+	 */
+	public String buildSentence(String bigram)
+	{
+		String out = "";
+		String s  = bigram.split(" ")[1];
+		String ss = nextWord(bigram);
+		out = bigram;
+		for (int i = 0; i< 13; i++)
+		{
+			String t = "";
+			out = out + " " + ss;
+			t = nextWord(s + " " +ss);
+			s = ss;
+			ss = t;
+		}
+		return out;
+	}
+	
+	
+	/**
+	 * get probability of triplet or a bigram
+	 * @param s
+	 * @return
+	 */
+	public double sP(String s)
+	{
+		double ret = 0d;
+		//System.out.println(s);
+		if (s.split(" ").length == 2)
+		{
+			//System.out.println(bigrams.get(s).doubleValue() + " / " + (double)bigramTotal);
+			try{ ret = bigrams.get(s).doubleValue() / (double)bigramTotal;}
+			catch(Exception e){ret = 0d;}
+		}
+		else if (s.split(" ").length == 3)
+		{
+			//System.out.println(trigrams.get(s).doubleValue() + " / " + (double)trigramTotal);
+			try{ ret = trigrams.get(s).doubleValue() / (double)trigramTotal;}
+			catch(Exception e){ret = 0d;}
+		}
+
+		return ret;
+	}
+	
+	/**
+	 * p(A | B C) = P (B C A) / P (B C).
+	 * compute P( B C A) and P( B C) 
+	 * frequency count for BCA and for BC, from the hashtables, and a count
+	 * for the total number of bigrams and trigrams.
+	 * P(B C A) should be frequence (B C A) / number of trigrams.
+	 * @return p(A | B C) = P (B C A) / P (B C)
+	 */
+	public double bigramCondProb(String A, String BC)
+	{
+		double bca = sP(BC+" "+A);
+		double bc = sP(BC);
+		if (bc==0)
+			return 0;
+		else
+			return bca / bc;
+	}
+	
+	
+	/**
+	 * sorts the dictionary
+	 * @param dict
+	 */
+	private void sortDict(HashMap<String, Integer> dict)
+	{	
+		for (String s: dict.keySet())
+		{
+			sortedDict.add(new sortString(s, dict.get(s)));
+		}
+	}
+	
+	/**
+	 * adds the string to the dictionary of words
+	 * @param one
+	 */
+	private void addToDict(HashMap<String, Integer> dict,String one)
+	{
+		if (dict.containsKey(one)) //update dict with new count
+		{	
+			int nc = dict.get(one);
+			nc++;
+			dict.put(one, nc);
+		}
+		else
+		{
+			dict.put(one, 1);
+		}
+
+	}
+	
+	/**
+	 *class to allow sorting of strings according to the count
+	 */
+	private class sortString implements Comparable<sortString>
+	{
+		public String string;
+		public int count;
+
+		sortString(String string, int count)
+		{
+			this.string = string;
+			this.count = count;
+		}
+		@Override
+		public int compareTo(sortString o)
+		{
+			int diff = o.count - this.count; //gross hack but works, reverse order
+			if (diff == 0)					 // breaks ties so that we don't lose fake dups by count
+				return 1;
+			else
+				return diff;
+		}
+		public String toString()
+		{
+			return string + "=" + count;
+		}
+		@Override
+		public boolean equals(Object o)
+		{
+			sortString s = (sortString) o;
+			return this.string == s.string;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.